simd/src/Kokkos_SIMD_Common.hpp

```diff
@@ -323,6 +331,43 @@ template <class T, class Abi>
   return a == simd_mask<T, Abi>(false);
 }
 
+template <class T, class Abi>
+[[nodiscard]] KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION T
+hmin(const_where_expression<simd_mask<T, Abi>, simd<T, Abi>> const& x) {
+  auto const& v = x.value();
+  auto const& m = x.mask();
+  auto result   = Kokkos::reduction_identity<T>::min();
+  for (std::size_t i = 0; i < v.size(); ++i) {
+    if (m[i]) result = Kokkos::min(result, v[i]);
+  }
+  return result;
+}
+
+template <class T, class Abi>
+[[nodiscard]] KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION T
+hmax(const_where_expression<simd_mask<T, Abi>, simd<T, Abi>> const& x) {
+  auto const& v = x.value();
+  auto const& m = x.mask();
+  auto result   = Kokkos::reduction_identity<T>::max();
+  for (std::size_t i = 0; i < v.size(); ++i) {
+    if (m[i]) result = Kokkos::max(result, v[i]);
+  }
+  return result;
+}
+
+template <class T, class Abi>
+[[nodiscard]] KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION T
+reduce(const_where_expression<simd_mask<T, Abi>, simd<T, Abi>> const& x, T,
```

Comment:
Where did the choice of the name `reduce` come from? This would seem to be more sensibly `hsum` (I'm assuming the `h` prefix is 'horizontal' across the elements of the vector)

Reply:
I see that's a pre-existing name in `Kokkos_SIMD_Scalar.hpp`. This is still in `namespace Experimental`, so what do others think about renaming this?

Reply:
@PhilMiller it comes from ISO C++ document N4808, the parallelism TS, which in turn is trying to be somewhat consistent with how ISO C++ reduction already look today. The inconsistency you're pointing out can be traced back to existing ISO C++ parallel algorithms, where there is nothing called `sum`, instead one uses `reduce` with `plus` as the reduction binary operation. Similarly, there is no binary operation usable for computing the minimum using `reduce` in ISO C++, instead one needs to use the `min` overload that takes a range. In my opinion, `min`, `max`, `hmin`, and `hmax` should all use `reduce` instead, but the design objective with Kokkos SIMD is to be consistent with the parallelism TS.

Reply:
I'm okay with defining an `hsum` that is only available in Kokkos, @crtrott and other can probably comment on the overall design and whether that is something we could submit as an ISO C++ paper or something.

Reply:
Ok, I hadn't realized the context of what we were aligning the design with. A possible amendment is worth talking about briefly, but it definitely doesn't play a role for this PR.